pkg/urbit/vere/io/ames.c

```diff
@@ -324,144 +325,143 @@ _ames_ship_of_chubs(c3_d sip_d[2], c3_y len_y, c3_y* buf_y)
+{
+  c3_y sen_y, rec_y;
+  c3_w cur_w = 0;
+
+  pre_u->sic_y = buf_y[0] & 0xf;
+  pre_u->ric_y = buf_y[0] & 0xf0;
+  cur_w++;
 
-  _ames_ship_to_chubs(pre_u->sen_d, sen_y, pre_y + 1);
-  _ames_ship_to_chubs(pre_u->rec_d, rec_y, pre_y + 1 + sen_y);
+  sen_y = 2 << hed_u->sac_y;
+  _ames_ship_to_chubs(pre_u->sen_d, sen_y, buf_y + cur_w);
+  cur_w += sen_y;
 
-  if (rog_y) {
+  rec_y = 2 << hed_u->rac_y;
+  _ames_ship_to_chubs(pre_u->rec_d, rec_y, buf_y + cur_w);
+  cur_w += rec_y;
+
+  if ( c3y == hed_u->rel_o ) {
     c3_y rag_y[8] = {0};
-    memcpy(rag_y, pre_y + 1 + sen_y + rec_y, rog_y);
+    memcpy(rag_y, buf_y + cur_w, 6);
     pre_u->rog_d = _ames_chub_bytes(rag_y);
   }
   else {
     pre_u->rog_d = 0;
   }
-  return len_y;
 }
 
+static c3_o
+_fine_sift_keen(u3_keen* ken_u, c3_w len_y, c3_y* buf_y)
+{
+  ken_u->fra_w = _ames_sift_word(buf_y);
+  ken_u->len_s = _ames_sift_short(buf_y + 4);
+  if ( ken_u->len_s > 384 ) {
+    return c3n;
+  }
+
+  ken_u->pat_c = c3_calloc(ken_u->len_s + 1);
+  memcpy(ken_u->pat_c, buf_y + 6, ken_u->len_s);
+  return c3y;
+}
 
 
-/* _fine_sift_requ(): parse request body, returning success
+/* _fine_sift_wail(): parse request body, returning success
 */
 static c3_o
-_fine_sift_requ(u3_head* hed_u,
-                u3_requ* req_u,
-                c3_w     len_w,
-                c3_y*    req_y)
+_fine_sift_wail(u3_pact* pac_u, c3_w cur_w)
 {
-  c3_y pre_y = _ames_sift_prelude(hed_u, &req_u->pre_u, len_w, req_y);
+  c3_w tot_w;
+  c3_s len_s;
 
-  req_y += pre_y;
-
-  c3_w rem_w = (64 + 4 + 2 + pre_y);
-  if ( rem_w > len_w ) {
-    u3l_log("not big enough\n");
+  if ( (cur_w + 64 + 4 + 2) > pac_u->len_w ) {
+    u3l_log("fine: wail not big enough\n");
     return c3n;
   }
-  memcpy(req_u->sig_y, req_y, 64);
-  req_y += 64;
+  memcpy(pac_u->wal_u.sig_y, pac_u->hun_y + cur_w, 64);
+  cur_w += 64;
 
-  req_u->fra_w = (
-      (req_y[0] << 0x0)
-    | (req_y[1] << 0x8)
-    | (req_y[2] << 0x10)
-    | (req_y[3] << 0x18));
-  req_y += 4;
+  pac_u->wal_u.ken_u.fra_w = _ames_sift_word(pac_u->hun_y + cur_w);
+  cur_w += 4;
 
-  req_u->len_s = c3_min(384,
-      (req_y[0] << 0x0)
-    | (req_y[1] << 0x8));
-  req_y += 2;
+  len_s = _ames_sift_short(pac_u->hun_y + cur_w);
+  pac_u->wal_u.ken_u.len_s = len_s;
+  cur_w += 2;
 
-  c3_w exp_w = rem_w + req_u->len_s;
-  if ( exp_w != len_w ) {
-    u3l_log("expected len: %u, actual %u\n", exp_w, len_w);
+  if ( len_s > 384 ) {
+    u3l_log("ames wail len: %u, max %u\n", len_s, 384);
     return c3n;
   }
 
-  req_u->pat_c = c3_calloc(req_u->len_s + 1);
+  tot_w = cur_w + len_s;
+  if ( tot_w != pac_u->len_w ) {
+    u3l_log("fine: wail expected total len: %u, actual %u\n",
+            tot_w, pac_u->len_w);
+    return c3n;
+  }
 
-  memcpy(req_u->pat_c, req_y, req_u->len_s);
-  req_u->pat_c[req_u->len_s] = '\0';
+  pac_u->wal_u.ken_u.pat_c = c3_calloc(len_s + 1);
+  memcpy(pac_u->wal_u.ken_u.pat_c, pac_u->hun_y + cur_w, len_s);
+  pac_u->wal_u.ken_u.pat_c[len_s] = '\0';
```

Comment:
A little bit surprised to see the null byte here. (Again, I realize it was in the old code too.) We know the length of the path string, right? Why do we need this to be a properly terminated C string? We end up needing to remember to `len_s + 1` whenever we allocate for it, which seems easy to overlook...

Reply:
it's generally nice to "over-terminate" when it's not too much trouble. even when the length is known, terminated strings are easier to print, play nicer with debuggers, &c.